lib/libvfio-user.c

```diff
@@ -188,17 +189,163 @@ dump_buffer(const char *prefix, const char *buf, uint32_t count)
     if (i % bytes_per_line != 0) {
         fprintf(stderr, "\n");
     }
-}
-#else
-#define dump_buffer(prefix, buf, count)
 #endif
+}
 
 static bool
 is_migr_reg(vfu_ctx_t *vfu_ctx, int index)
 {
     return &vfu_ctx->reg_info[index] == vfu_ctx->migr_reg;
 }
 
+static ssize_t
+region_access(vfu_ctx_t *vfu_ctx, size_t region_index,
+              bool is_write, char *buf, uint64_t offset, size_t count)
+{
+    ssize_t ret;
+
+    assert(vfu_ctx != NULL);
+    assert(buf != NULL);
+
+    vfu_log(vfu_ctx, LOG_DEBUG, "%s %zu %#lx-%#lx", is_write ? "W" : "R",
+            region_index, offset, offset + count);
+
+    if (is_write) {
+        dump_buffer("buffer write", buf, count);
+    }
+
+    if (region_index == VFU_PCI_DEV_CFG_REGION_IDX) {
+        ret = pci_config_space_access(vfu_ctx, buf, count, offset, is_write);
+    } else if (is_migr_reg(vfu_ctx, region_index)) {
+        ret = migration_region_access(vfu_ctx, buf, count, offset, is_write);
+    } else {
+        vfu_region_access_cb_t *cb = vfu_ctx->reg_info[region_index].fn;
+
+        if (cb == NULL) {
+            vfu_log(vfu_ctx, LOG_ERR, "no callback for region %d",
+                    region_index);
+            return -EINVAL;
+        }
+
+        ret = cb(vfu_ctx, buf, count, offset, is_write);
+    }
+
+    if (!is_write && (size_t)ret == count) {
+        dump_buffer("buffer read", buf, count);
+    }
+
+    return ret;
+}
+
+static bool
+valid_region_access(vfu_ctx_t *vfu_ctx, size_t size, uint16_t cmd,
+                    struct vfio_user_region_access *ra)
+{
+    size_t index;
+
+    assert(ra != NULL);
```

Comment:
We should validate `vfu_ctx` as well (not done in the original code).

Reply:
done